macho.go

```diff
@@ -34,14 +36,70 @@ func openMachO(fp string) (*machoFile, error) {
 	if err != nil {
 		return nil, fmt.Errorf("error when parsing the Mach-O file: %w", err)
 	}
-	return &machoFile{file: f, osFile: osFile}, nil
+	return &machoFile{file: f, osFile: osFile, symtab: newSymbolTableOnce()}, nil
 }
 
 var _ fileHandler = (*machoFile)(nil)
 
 type machoFile struct {
 	file   *macho.File
 	osFile *os.File
+	symtab *symbolTableOnce
+}
+
+func (m *machoFile) initSymtab() error {
+	m.symtab.Do(func() {
+		if m.file.Symtab == nil {
+			// just do nothing, keep err nil and table empty
+			return
+		}
+		const stabTypeMask = 0xe0
+		// Build a sorted list of addresses of all symbols.
+		// We infer the size of a symbol by looking at where the next symbol begins.
+		var addrs []uint64
+		for _, s := range m.file.Symtab.Syms {
+			// Skip stab debug info.
+			if s.Type&stabTypeMask == 0 {
+				addrs = append(addrs, s.Value)
+			}
+		}
+		slices.Sort(addrs)
+
+		var syms []symbol
+		for _, s := range m.file.Symtab.Syms {
+			if s.Type&stabTypeMask != 0 {
+				// Skip stab debug info.
+				continue
+			}
+			sym := symbol{Name: s.Name, Value: s.Value}
+			i := sort.Search(len(addrs), func(x int) bool { return addrs[x] > s.Value })
+			if i < len(addrs) {
+				sym.Size = addrs[i] - s.Value
+			}
+			syms = append(syms, sym)
+		}
+
+		for _, sym := range syms {
+			m.symtab.table[sym.Name] = sym
+		}
+	})
+	return m.symtab.err
+}
+
+func (m *machoFile) hasSymbolTable() (bool, error) {
+	return m.file.Symtab != nil, nil
```

Comment:
Since `newSymbolTableOnce()` is always called first, this will always be true. Do we have a need for this function in other cases when it's not called before a `getSymbol` call? Better approach is to handle the check in `getSymbol` and return an error type when the table doesn't exist.

Reply:
This judgment can be expensive on some platforms, such as on elf, where determining this requires a large number of memory copies. One such function allows us to skip certain steps with O(1) complexity.

Reply:
Resolved by sync.Value